core/src/main/java/org/jdbi/v3/core/ReduceRow.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core;
+
+import java.lang.reflect.Type;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+import org.jdbi.v3.core.exception.UnableToExecuteStatementException;
+
+/**
+ * A ReduceRow is a view over a single row of a {@code ResultSet}
+ * that uses the current statement context to map rows and
+ * columns to values.  It is not valid past the return of
+ * the method that receives it.
+ */
+public class ReduceRow
+{
+    private final StatementContext ctx;
+    private final ResultSet rs;
+
+    ReduceRow(ResultSet rs, StatementContext ctx)
+    {
+        this.rs = rs;
+        this.ctx = ctx;
+    }
+
+    /**
+     * Use a row mapper to extract a type from the current ResultSet row.
+     */
+    public <T> T getRow(Class<T> rowClass)
```

Comment:
If we made JoinRow look like ReduceRow (lazy mapping, map row or column types, no need to register types joined on row) then I see a very strong case for JoinRow being the default mapped typed instead of Map<String,Object>